router/h2/src/main/scala/com/twitter/finagle/buoyant/h2/H2FailureAccrualFactory.scala

```diff
@@ -0,0 +1,254 @@
+    state match {
+      case ProbeOpen =>
+        probesCounter.incr()
+        state = ProbeClosed
+      case _ =>
+    }
+  }
+
+  private[this] def classifyResponse(classifier: H2Classifier, h2ReqRep: H2ReqRep): Try[Response] =
+    classifier.responseClassifier.lift(h2ReqRep) match {
+      case Some(ResponseClass.Successful(_)) =>
+        didSucceed()
+        h2ReqRep.response
+      case Some(ResponseClass.Failed(_)) =>
+        didFail()
+        h2ReqRep.response
+      case None =>
+        h2ReqRep.response match {
+          case Return(rep) =>
+            if (rep.stream.isEmpty) {
+              classifyStream(classifier, H2ReqRepFrame(h2ReqRep.request, Return((rep, None))))
+              Return(rep)
+            } else {
+              val stream = rep.stream.onFrame {
+                case Return(f) if f.isEnd =>
+                  classifyStream(classifier, H2ReqRepFrame(h2ReqRep.request, Return((rep, Some(Return(f))))))
+                case Throw(e) =>
+                  classifyStream(classifier, H2ReqRepFrame(h2ReqRep.request, Return((rep, Some(Throw(e))))))
+                case _ => ()
+              }
+              Return(Response(rep.headers, stream))
+            }
+          case Throw(e) =>
+            classifyStream(classifier, H2ReqRepFrame(h2ReqRep.request, Throw(e)))
+            Throw(e)
+        }
+    }
+
+  private[this] def classifyStream(classifier: H2Classifier, h2ReqRepFrame: H2ReqRepFrame): Unit =
+    classifier.streamClassifier(h2ReqRepFrame) match {
+      case ResponseClass.Successful(_) => didSucceed()
+      case ResponseClass.Failed(_) => didFail()
+    }
+
+  def apply(conn: ClientConnection) = {
+    underlying(conn).map { service =>
+      // N.B. the reason we can't simply filter the service factory is so that
+      // we can override the session status to reflect the broader endpoint status.
+      new Service[Request, Response] {
+        def apply(request: Request): Future[Response] = {
+          val param.H2Classifier(classifier) =
+            H2ClassifierCtx.current.getOrElse(param.H2Classifier.param.default)
+          // If service has just been revived, accept no further requests.
+          // Note: Another request may have come in before state transitions to
+          // ProbeClosed, so > 1 requests may be processing while in the
+          // ProbeClosed state. The result of first to complete will determine
+          // whether the factory transitions to Alive (successful) or Dead
+          // (unsuccessful).
+          stopProbing()
+
+          service(request).transform { rep =>
+            Future.const(classifyResponse(classifier, H2ReqRep(request, rep)))
+          }
+        }
+
+        override def close(deadline: Time): Future[Unit] = service.close(deadline)
+        override def status: FStatus = FStatus.worst(
+          service.status,
+          H2FailureAccrualFactory.this.status
+        )
+      }
+    }.onFailure(onServiceAcquisitionFailure)
```

Comment:
Instead of splitting factory into a val, I changed the Service into a named inner class.  Let me know if you think this improves readability.